src/fusion_algorithm.cpp

```diff
@@ -0,0 +1,325 @@
+
+    for (int i = 0; i < number_of_sensors; i++) {
+        for (int j = 0; j < number_of_sensors; j++) {
+            gsl_matrix_set(p_eigen_vec, i, j, eignvectors[i][j]);
+            gsl_matrix_set(p_support_degree, i, j,
+                           degree_matrix[j + (i * number_of_sensors)]);
+            gsl_matrix_set(p_y, i, j, 0.0);
+        }
+    }
+
+
+    gsl_blas_dgemm(CblasNoTrans, CblasNoTrans,
+                   1.0, p_eigen_vec, p_support_degree,
+                   0.0, p_y);
+
+    printf("\n\nPrincipal components : \n");
+    for (int rows = 0; rows < number_of_sensors; rows++) {
+
+        for (int cols = 0; cols < number_of_sensors; cols++) {
+            principal_components[rows][cols] = gsl_matrix_get(p_y, rows, cols);
+            printf("%f, ", principal_components[rows][cols]);
+        }
+    }
+    printf("\n");
+    return principal_components;
+}
+
+double *get_contribution_rates(double *eigenvalues, int number_of_sensors) {
+    double *contribution_rates = (double *) calloc(number_of_sensors, sizeof(double));
+    double eignvalue_sum = 0.0;
+    for (int i = 0; i < number_of_sensors; ++i) {
+        double val = eigenvalues[i];
+        eignvalue_sum += eigenvalues[i];
+    }
+    printf("\n\nContribution rates: :\n");
+
+    for (int i = 0; i < number_of_sensors; ++i) {
+        contribution_rates[i] = eigenvalues[i] / eignvalue_sum;
+        printf("%f, ", contribution_rates[i]);
+    }
+    printf("\n");
+    return contribution_rates;
+}
+
+
+int select_contribution_rate(double *contribution_rates, int number_of_sensors,
+                             float p) {
+    double *cumulative_contributions = (double *) malloc(number_of_sensors * sizeof(double));
+
+    for (int i = 0; i < number_of_sensors; ++i) {
+        cumulative_contributions[i] = 0.0;
+        for (int j = 0; j <= i; ++j) {
+            cumulative_contributions[i] += contribution_rates[j];
+        }
+    }
+    printf("\n\nStep 5: \n");
+    for (int k = 0; k < number_of_sensors; ++k) {
+        if (cumulative_contributions[k] > p) {
+            int m = k + 1;
+            printf("The value of m is: %d\n", m);
+            return m;
+        }
+    }
+
+}
+
+double *get_integrated_support_scores(double **principal_components,
+                              double *contribution_rates, int number_of_sensors, int m) {
+    double *integrated_scores = (double *) calloc(number_of_sensors, sizeof(double));
+
+    for (int i = 0; i < number_of_sensors; i++) {
+        for (int j = 0; j < m; j++) {
+            integrated_scores[i] += (contribution_rates[j] *principal_components[j][i]);
+        }
+    }
+    printf("Step 6\n Integrated scores: [");
+
+    for (int j = 0; j < number_of_sensors; ++j) {
+        printf("%f, ", integrated_scores[j]);
+    }
+    printf("]\n");
+    return integrated_scores;
+}
+
+
+SensorsList_t eliminate_incorrect_data(SensorsList_t sensors_list, double *integrated_scores, float tolerance) {
```

Comment:
Extremely heavy action. Copying entire SensorList, adjusting. Then sending a copy back. Also, the implementation is not quite right. Some unnecessary use of dynamic memory allocations. will adjust

Reply:
Vectors have methods to delete elements. They are not like arrays. It is not needed to loop through and adjust each element. and Push back. A simple loop to check if incorrect and an .erase method.